walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,175 @@
+                            optimalAvailableTimes.put(mandatoryAttendeesAvailableTime, count + 1);
+                        } else {
+                            optimalAvailableTimes.put(mandatoryAttendeesAvailableTime, 1);
+                        }
+                    }
+                }
+            }
+        }
+        int maximum = 0;
+        ArrayList<TimeRange> optimalAvailableTime = new ArrayList<TimeRange>();
+        for (Map.Entry<TimeRange, Integer> entry : optimalAvailableTimes.entrySet()) {
+            TimeRange availableTime = entry.getKey();
+            int count = entry.getValue();
+            if (count > maximum) {
+                maximum = count;
+                optimalAvailableTime.clear();
+                optimalAvailableTime.add(availableTime);
+            } else if (count == maximum) {
+                optimalAvailableTime.add(availableTime);
+            }
+        }
+        if (optionalAttendeesAvailableTimes.size() > 0) {
+            return optionalAttendeesAvailableTimes;
+        } else {
+            if (optimalAvailableTime.size() > 0) {
+                return optimalAvailableTime;
+            } else {
+                return mandatoryAttendeesAvailableTimes;
+            }
+        }
+    }
+
+    private ArrayList<TimeRange> findAvailableTimes(ArrayList<TimeRange> unavailableTimes, MeetingRequest request) {
+        // Sort time ranges by start time.
+        TimeRangeComparator comparator = new TimeRangeComparator();
+        unavailableTimes.sort(comparator);
+        // Combine overlapping time ranges.
+        int i = 0;
+        while (i < unavailableTimes.size() - 1) {
+            if (unavailableTimes.get(i).overlaps(unavailableTimes.get(i + 1)) == true) {
+                int start = unavailableTimes.get(i).start();
+                int end;
+                if (Long.compare(unavailableTimes.get(i).end(), unavailableTimes.get(i + 1).end()) > 0) {
+                    end = unavailableTimes.get(i).end();
+                } else {
+                    end = unavailableTimes.get(i + 1).end();
+                }
+                unavailableTimes.set(i, TimeRange.fromStartEnd(start, end, false));
+                unavailableTimes.remove(i + 1);
+            } else {
+                i++;
+            }
+        }
+
+        ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>();
+        if (unavailableTimes.size() == 0) {
+            // If there are no time slots that attendees are unavailable, the whole day is available.
+            int start;
+            int end;
+            TimeRange time;
+            start = TimeRange.START_OF_DAY;
+            end = TimeRange.END_OF_DAY;
+            time = TimeRange.fromStartEnd(start, end, true);
+            if (time.duration() >= request.getDuration()) {
+                availableTimes.add(time);
+            }
+        } else if (unavailableTimes.size() == 1) {
```

Comment:
I'm a little surprised you have to handle a single unavailable time specially.  Could that just be handled by the code starting on line 157?